### PR TITLE
initial pass implementation of OIDC

### PR DESCRIPTION
**Pull request by andrewroth**
this is from some back and forth with claude code

### CODE REVIEW

File: app/controllers/auth_controller.rb [hunk 1]
```diff
@@ -20,8 +20,16 @@ def create
...
       user = GoogleAuthService.find_user_by_token(data_attrs[:google_id_token], data_attrs[:create_user])
       AuthToken.new(user: user)
-    when :okta, :facebook
-      user = "::#{method.to_s.capitalize}AuthService".constantize.find_user_by_token(data_attrs[:"#{method}_access_token"], data_attrs[:create_user])
+    when :okta
+      user = OktaAuthService.find_user_by_token(data_attrs[:okta_access_token], data_attrs[:create_user])
+      AuthToken.new(user: user)
+    when :facebook
+      # Handle both access tokens and ID tokens for Facebook
+      if data_attrs[:facebook_access_token]
```
Comment (knutsenm) [inline app/controllers/auth_controller.rb:28]: This appears to be the old method you're replacing. Shouldn't a check for the new method (line 30) come first?

**Review by knutsenm — CHANGES_REQUESTED** [review]

File: app/controllers/auth_controller.rb (hunk 1 above)
Comment (frett) [reply app/controllers/auth_controller.rb:28]: I'm indifferent as to the order of these, we should only be providing/validating one anyways.

I feel if a client passes multiple auth credentials it should do one of these things:
- attempt to validate one, and ignore the others. The one it chooses to validate should be considered undefined behavior.
- return an error about providing multiple credentials

Comment (frett) [reply app/controllers/auth_controller.rb:28]: also, `facebook_access_token` isn't an old method currently, it is still the auth mechanism for Facebook login on the Android App. 

We had to make the change for iOS because Apple/Facebook were forcing an additional user confirmation prompt when using a facebook_access_token in the iOS app. If a similar problem shows up in Android we can switch, but there are no intentions of switching the Android mechanism at this time.

Comment (dr-bizz) [inline app/controllers/auth_controller.rb:28]: [Must-fix] Missing validation when both facebook_access_token and facebook_id_token are provided. Evidence: Lines 28-32 allow both tokens to be processed. Impact: Could lead to AuthToken.new(user: nil) causing downstream issues. Fix: Add mutual exclusion validation. [Category: File]

File: app/controllers/auth_controller.rb [hunk 2]
```diff
@@ -20,8 +20,16 @@ def create
...
+        user = FacebookAuthService.find_user_by_token(data_attrs[:facebook_access_token], data_attrs[:create_user])
+      elsif data_attrs[:facebook_id_token]
+        user = FacebookOidcAuthService.find_user_by_token(data_attrs[:facebook_id_token], data_attrs[:create_user])
+      end
```
Comment (dr-bizz) [inline app/controllers/auth_controller.rb:32]: [Nice-to-have] Consider explicit error when neither facebook token is provided. Evidence: Lines 28-32 allow user=nil scenario. Impact: Silent failures harder to debug. Suggestion: Add validation for missing tokens. [Category: File]

File: app/services/facebook_oidc_auth_service.rb [hunk 3]
```diff
@@ -0,0 +1,92 @@
...
+      user_attributes[:last_name] = decoded_token["family_name"] if decoded_token["family_name"]
+
+      # Set facebook_user_id to maintain consistency with existing users
+      user_attributes[:facebook_user_id] = decoded_token["user_id"]
+
+      user_attributes.with_indifferent_access
+    end
+
+    def fetch_facebook_jwks
+      @facebook_jwks ||= begin
```
Comment (dr-bizz) [inline app/services/facebook_oidc_auth_service.rb:74]: [Must-fix] JWKS cache has no expiration mechanism. Evidence: @facebook_jwks ||= begin at line 74. Impact: Stale keys could cause auth failures until restart. Fix: Add time-based cache expiration. [Category: File]

File: app/services/facebook_oidc_auth_service.rb [hunk 4]
```diff
@@ -0,0 +1,92 @@
...
+    def decode_token(id_token)
+      jwks = fetch_facebook_jwks
+      jwt_payload = JWT.decode(id_token, nil, true, {
+        algorithms: ["RS256"],
+        jwks: jwks,
+        verify_aud: true,
+        aud: ENV.fetch("FACEBOOK_APP_ID"),
+        verify_iss: true,
+        iss: "https://www.facebook.com"
+      })[0]
```
Comment (dr-bizz) [inline app/services/facebook_oidc_auth_service.rb:27]: [Nice-to-have] JWT validation options duplicated in tests. Evidence: Similar JWT.decode options in lines 19-27. Impact: Test maintenance burden. Suggestion: Extract to shared configuration. [Category: File]

File: spec/acceptance/auth_controller_spec.rb [hunk 5]
```diff
@@ -265,6 +265,173 @@
...
+
+          # Mock the JWKS fetching
+          allow(Net::HTTP).to receive(:get_response).with(URI("https://limited.facebook.com/.well-known/openid-configuration/"))
+            .and_return(double("response", is_a?: Net::HTTPSuccess, body: '{"jwks_uri": "https://www.facebook.com/.well-known/oauth/jwks"}'))
+
+          allow(Net::HTTP).to receive(:get_response).with(URI("https://www.facebook.com/.well-known/oauth/jwks"))
+            .and_return(double("response", is_a?: Net::HTTPSuccess, body: '{"keys": [{"kty": "RSA", "kid": "test"}]}'))
+
+          # Now create a specific mock for our Facebook OIDC token
+          facebook_jwt_options = {
```
Comment (dr-bizz) [inline spec/acceptance/auth_controller_spec.rb:284]: [Nice-to-have] facebook_jwt_options hash duplicated across multiple test cases. Evidence: Lines 284-291, 345-352, 366-373, 387-394, 408-415. Impact: Code duplication increases maintenance. Suggestion: Extract to let block. [Category: File]

File: app/controllers/auth_controller.rb (hunk 1 above)
Comment (dr-bizz) [inline app/controllers/auth_controller.rb:28]: [Suggested tests] Add test cases for edge scenarios: both facebook tokens provided, neither provided. Evidence: Current tests only cover single token scenarios. Impact: Missing validation coverage. [Category: File]

File: app/services/facebook_oidc_auth_service.rb [hunk 6]
```diff
@@ -0,0 +1,92 @@
...
+    def primary_key
+      :facebook_user_id
+    end
+
+    def expected_fields
+      %w[user_id]
+    end
+
+    def decode_token(id_token)
+      jwks = fetch_facebook_jwks
```
Comment (dr-bizz) [inline app/services/facebook_oidc_auth_service.rb:19]: [Reuse] Could extract JWT validation logic to BaseAuthService. Evidence: Similar JWT.decode patterns in OktaAuthService line 15. Impact: Standardizes OIDC validation across services. [Category: Reuse]

File: app/services/facebook_oidc_auth_service.rb [hunk 7]
```diff
@@ -0,0 +1,92 @@
...
+      # Only set attributes if they exist in the token
+      user_attributes[:email] = decoded_token["email"] if decoded_token["email"]
+      user_attributes[:name] = decoded_token["name"] if decoded_token["name"]
+      user_attributes[:first_name] = decoded_token["given_name"] if decoded_token["given_name"]
+      user_attributes[:last_name] = decoded_token["family_name"] if decoded_token["family_name"]
+
+      # Set facebook_user_id to maintain consistency with existing users
+      user_attributes[:facebook_user_id] = decoded_token["user_id"]
+
+      user_attributes.with_indifferent_access
```
Comment (dr-bizz) [inline app/services/facebook_oidc_auth_service.rb:70]: [Mentoring] Good separation of concerns following BaseAuthService pattern. The user attribute mapping could be simplified using a hash-based approach for consistency. [Category: File]

**Review by dr-bizz — COMMENTED** [review]
This is an AI-generated review. (Mark asked me to post)

----
Adds Facebook OIDC authentication support with comprehensive JWT validation. The implementation follows existing patterns well but needs input validation in the controller and JWKS cache expiration. Security enhancements to existing Facebook auth service are well-implemented.

File: app/controllers/auth_controller.rb (hunk 1 above)
Comment (frett) [reply app/controllers/auth_controller.rb:28]: I would actually say a better way to refactor this is to get rid of the extraction of the credential handler, and just do a big switch based upon which credential is in the request. that way we don't need to have a special case for facebook when the already existing default fallback of code could handle it just fine.

File: app/services/facebook_oidc_auth_service.rb (hunk 4 above)
Comment (knutsenm) [reply app/services/facebook_oidc_auth_service.rb:27]: Closing; I believe we don't care if tests are WET

File: app/controllers/auth_controller.rb (hunk 1 above)
Comment (knutsenm) [reply app/controllers/auth_controller.rb:28]: It's not a big deal, but I feel that placing the `facebook_id_token` check first would indicate it's the preferred, modern method.

**Review by knutsenm — APPROVED** [review]
🎉
